feat: normalize `packageManager` field when parsing

BobbieGoede opened the pull request:
Resolves #176 

This removes the pattern `^\W+` to drop unexpected characters preceding the package manager name (and version), while still allowing unknown name patterns `\w+`. 

If the pattern removal results in a different package name a warning will be returned and logged when used in the CLI detect command, the warning can be disabled by passing the `--warn false`.

### Notes

* The `sanitizePackageManagerName` returns `any` for the sanitized name, this matches the original behavior, but gives a false sense of type safety, perhaps something to look into in a separate PR.

BobbieGoede commented:
I'll look into adding a simple unit test for the function

pi0 commented on src/package-manager.ts:
```diff
@@ -70,6 +78,19 @@ export const packageManagers: PackageManager[] = [
+  const sanitized = name.replace(/^\W+/, "");
+  if (name !== sanitized) {
+    if (!disableWarn) {
+      consola.warn(
```
API utils don't show warning (or depend on consola) we might return an object and pass it to final resolve object (something like `abnormalChars: true` or `warnings: []` but no console warn here)

BobbieGoede commented:
@pi0 
I have renamed the function and moved it to `_utils.ts` to keep it encapsulated while being able to import it for tests and added basic tests.

Warnings are now added to an optional `warnings` property on `PackageManager`, maybe a bit odd for one warning but will be less intrusive if more warnings are added in the future.

pi0 commented on src/_utils.ts:
```diff
@@ -158,3 +158,16 @@ export function doesDependencyExist(
+  name: string,
+): [any, string | undefined] {
+  const sanitized = name.replace(/^\W+/, "");
+  if (name !== sanitized) {
```
Wondering why not simply using a regex test /^\w+$`

BobbieGoede replied:
Ah, at first I thought we would be sanitizing the full `packageManager` string which potentially includes the version, but since the version is already split we could test/exec for just the word characters.

pi0 replied:
you are right i forgot that context.

pi0 commented on src/_utils.ts:
```diff
@@ -158,3 +158,16 @@ export function doesDependencyExist(
+      `Abnormal characters found in \`packageManager\` field, sanitizing from \`'${name}'\` to \`'${sanitized}'\``,
+    ];
+  }
+  return [name, undefined];
```
Could be better for readability if we return `{ name, warnings }`

pi0 commented on src/cli.ts:
```diff
@@ -85,14 +85,27 @@ const detect = defineCommand({
     const cwd = resolve(args.cwd || ".");
     const packageManager = await detectPackageManager(cwd);
+
+    if (args.warn && packageManager?.warnings) {
```
I think warn by default in cli is fine. If someone wants to suppress warn/errors can do `2>dev/null`

pi0 commented on src/_utils.ts:
```diff
@@ -158,3 +158,19 @@ export function doesDependencyExist(
+    };
+  }
+
+  return { name, warning: undefined };
```
Returning `name: sanitized` was fine. For consistency we could use `warnings` btw.

pi0 commented on src/_utils.ts:
```diff
@@ -158,3 +158,19 @@ export function doesDependencyExist(
+  name: any;
+  warning: string | undefined;
+} {
+  const sanitized = /\w+$/.exec(name)?.[0];
```
Even though today we don't have any package manager with `-` in it, it could be possible.

BobbieGoede commented:
@pi0 
Hmm I don't quite follow, should we use the previous `/^\w+/` replacement instead of matching `/\w+$/`?

> For consistency we could use warnings btw.

So we want the pattern to be a `warnings` string array when applicable?

pi0 commented:
Thanks for the PR. worked on it a little bit more. If the name part matches the npm package name regex, we parse it as-is, otherwise, we fall back to sanitize+warn mode.

I also added buildMeta with the new parser util to be exposed.
